// angular-date-value-accessor/date-value-accessor.js
"use strict";
var core_1 = require('@angular/core');
var forms_1 = require('@angular/forms');
exports.DATE_VALUE_ACCESSOR = {
    provide: forms_1.NG_VALUE_ACCESSOR,
    useExisting: core_1.forwardRef(function () { return DateValueAccessor; }),
    multi: true
};
/**
 * The accessor for writing a value and listening to changes on a date input element
 *
 *  ### Example
 *  `<input type="date" name="myBirthday" ngModel useValueAsDate>`
 */
var DateValueAccessor = (function () {
    function DateValueAccessor(_renderer, _elementRef) {
        this._renderer = _renderer;
        this._elementRef = _elementRef;
        this.onChange = function (_) { };
        this.onTouched = function () { };
    }
    DateValueAccessor.prototype.writeValue = function (value) {
        if (!value) {
            this._renderer.setElementProperty(this._elementRef.nativeElement, 'value', null);
            return;
        }
        this._renderer.setElementProperty(this._elementRef.nativeElement, 'valueAsDate', value);
    };
    DateValueAccessor.prototype.registerOnChange = function (fn) { this.onChange = fn; };
    DateValueAccessor.prototype.registerOnTouched = function (fn) { this.onTouched = fn; };
    DateValueAccessor.prototype.setDisabledState = function (isDisabled) {
        this._renderer.setElementProperty(this._elementRef.nativeElement, 'disabled', isDisabled);
    };
    DateValueAccessor.decorators = [
        { type: core_1.Directive, args: [{
                    // this selector changes the previous behavior silently and might break existing code
                    // selector: 'input[type=date][formControlName],input[type=date][formControl],input[type=date][ngModel]',
                    // this selector is an opt-in version
                    selector: '[useValueAsDate]',
                    providers: [exports.DATE_VALUE_ACCESSOR]
                },] },
    ];
    /** @nocollapse */
    DateValueAccessor.ctorParameters = [
        { type: core_1.Renderer, },
        { type: core_1.ElementRef, },
    ];
    DateValueAccessor.propDecorators = {
        'onChange': [{ type: core_1.HostListener, args: ['input', ['$event.target.valueAsDate'],] },],
        'onTouched': [{ type: core_1.HostListener, args: ['blur', [],] },],
    };
    return DateValueAccessor;
}());
exports.DateValueAccessor = DateValueAccessor;
